Modules/EMCAL/src/ClusterTask.cxx

```diff
@@ -0,0 +1,217 @@
+
+  mHistNCells_EMCal = new TH2F("Cells_EMCal", "No of cells in a cluster;Cluster E;N^{EMC}_{cells}", 500, 0, 50, 30, 0, 30);
+  getObjectsManager()->startPublishing(mHistNCells_EMCal);
+
+  mHistM02_EMCal = new TH1F("M02_EMCal", "M02 distribution; M02;Counts", 200, 0.0, 2.0);
+  getObjectsManager()->startPublishing(mHistM02_EMCal);
+
+  mHistM20_EMCal = new TH1F("M20_EMCal", "M20 distribution; M20;Counts", 200, 0.0, 2.0);
+  getObjectsManager()->startPublishing(mHistM20_EMCal);
+
+  mHistClustE_DCal = new TH1F("ClustE_DCal", "Cluster energy distribution; Cluster E(GeV);Counts", 500, 0.0, 50.0);
+  getObjectsManager()->startPublishing(mHistClustE_DCal);
+
+  mHistNCells_DCal = new TH2F("Cells_DCal", "No of cells in a cluster;Cluster E;N^{EMC}_{cells}", 500, 0, 50, 30, 0, 30);
+  getObjectsManager()->startPublishing(mHistNCells_DCal);
+
+  mHistM02_DCal = new TH1F("M02_DCal", "M02 distribution; M02;Counts", 200, 0.0, 2.0);
+  getObjectsManager()->startPublishing(mHistM02_DCal);
+
+  mHistM20_DCal = new TH1F("M20_DCal", "M20 distribution; M20;Counts", 200, 0.0, 2.0);
+  getObjectsManager()->startPublishing(mHistM20_DCal);
+}
+
+void ClusterTask::startOfActivity(Activity& activity)
+{
+  ILOG(Info, Support) << "startOfActivity " << activity.mId << ENDM;
+  resetHistograms();
+}
+
+void ClusterTask::startOfCycle()
+{
+  ILOG(Info, Support) << "startOfCycle" << ENDM;
+  resetHistograms();
+}
+
+void ClusterTask::monitorData(o2::framework::ProcessingContext& ctx)
+{
+
+  std::string InputCell, InputCluster, InputCellIndex;
+  std::string InputCellTrig, InputClusterTrig, InputCellIndexTrig;
+
+  InputCell = "emcal-cells";
+  InputCellTrig = "emcal-cellstriggerecords";
+
+  InputCluster = "emcal-clusters";
+  InputClusterTrig = "emcal-clustertriggerecords";
+
+  InputCellIndex = "emcal-cellindices";
+  InputCellIndexTrig = "emcal-citriggerecords";
+
+  auto Cell = ctx.inputs().get<gsl::span<o2::emcal::Cell>>(InputCell.c_str());
+
+  auto CellTriggerRecord = ctx.inputs().get<gsl::span<o2::emcal::TriggerRecord>>(InputCellTrig.c_str());
+
+  auto Cluster = ctx.inputs().get<gsl::span<o2::emcal::Cluster>>(InputCluster.c_str());
+
+  auto ClusterTriggerRecord = ctx.inputs().get<gsl::span<o2::emcal::TriggerRecord>>(InputClusterTrig.c_str());
+
+  auto CellIndex = ctx.inputs().get<gsl::span<int>>(InputCellIndex.c_str());
+
+  auto CellIndexTriggerRecord = ctx.inputs().get<gsl::span<o2::emcal::TriggerRecord>>(InputCellIndexTrig.c_str());
+
+  mEventHandler = std::make_unique<o2::emcal::EventHandler<o2::emcal::Cell>>();
+  mClusterFactory = std::make_unique<o2::emcal::ClusterFactory<o2::emcal::Cell>>();
+
+  mEventHandler->setClusterData(Cluster, CellIndex, ClusterTriggerRecord, CellIndexTriggerRecord);
+  mEventHandler->setCellData(Cell, CellTriggerRecord);
+
+  for (int iev = 0; iev < mEventHandler->getNumberOfEvents(); iev++) {
+    auto inputEvent = mEventHandler->buildEvent(iev);
+
+    resetHistograms();
```

Comment:
Histograms must not be reset in the event loop. The idea is to fill histograms over a longer time in order to accumulate some statistics. Usually histograms a accumulative and are reset only at the start of run (in startOfActity), with an option to have this per cycle duration.

Furthermore: If you reset histograms within the event loop the data will never get published - there is no publishing of histograms while the run function is still active (no asynchronous processing). If you reset the histograms in the loop the data is lost forever.

Never call resetHistograms actively (except from startOfActivity) - let this be done by the framework.